In [1]:
import pandas as pd
import chainladder as cl

In [2]:
wiki = pd.read_csv('wikipedia.csv', sep=';')[['AccidentYear', 'DevelopmentYear', 'Reported', 'Paid']]

In [3]:
xyz_tri = cl.Triangle(
    data=wiki,
    origin="AccidentYear",
    development="DevelopmentYear",
    columns=["Reported", "Paid"],
    cumulative=True,
)
xyz_tri

Triangle Summary
Valuation:           2007-12
Grain:                  OYDY
Shape:        (1, 2, 10, 10)
Index:               [Total]
Columns:    [Reported, Paid]

### Age-to-age factors

In [4]:
xyz_tri['Reported'].age_to_age.to_frame().style.format(precision=3)

### Simple average last 5 years

In [5]:
cl.Development(average="simple", n_periods=5).fit(xyz_tri["Reported"]).ldf_.to_frame().style.format(precision=3)

### Simple average last 3 years

In [6]:
cl.Development(average="simple", n_periods=3).fit(xyz_tri["Reported"]).ldf_.to_frame().style.format(precision=3)

### Volume weighted last 5 years

In [7]:
cl.Development(average="volume", n_periods=5).fit(xyz_tri["Reported"]).ldf_.to_frame().style.format(precision=3)

### Volume weighted last 5 years

In [8]:
cl.Development(average="volume", n_periods=3).fit(xyz_tri["Reported"]).ldf_.to_frame().style.format(precision=3)

### Cumulative to ultimate

In [9]:
cl.Development(average="volume", n_periods=3).fit(xyz_tri["Reported"]).cdf_

12-Ult    24-Ult    36-Ult    48-Ult    60-Ult    72-Ult    84-Ult    96-Ult   108-Ult
(All)  1.289992  1.108105  1.049464  1.021522  1.009899  1.005305  1.002548  1.000949  1.000369

### Estimation of ultimate reported claims

In [21]:
xyz_dev = cl.Pipeline(
    [("dev", cl.Development(average="volume", n_periods=3))]
).fit_transform(xyz_tri['Reported'])
xyz_model = cl.Chainladder().fit(xyz_dev)
xyz_model.cdf_.to_frame().style.format(precision=3)

In [19]:
xyz_model.ultimate_

2261
1998  4.774230e+07
1999  5.120467e+07
2000  5.488995e+07
2001  5.644303e+07
2002  5.890357e+07
2003  5.813516e+07
2004  5.820291e+07
2005  5.959528e+07
2006  6.054836e+07
2007  6.302072e+07

#### Estimation of ultimate claims (from Wikipedia)
|Accident year	| Reported claims	| Development factor to ultimate	| Projected ultimate claims |
|---|---|---|---|
|1998	| 47,742,304	| 1.000	| 47,742,304 |
|1999	| 51,185,767	| 1.000	| 51,185,767 |
|2000	| 54,837,929	| 1.001	| 54,892,767 |
|2001	| 56,299,562	| 1.003	| 56,468,461 |
|2002	| 58,592,712	| 1.006	| 58,944,268 |
|2003	| 57,565,344	| 1.011	| 58,198,563 | 
|2004	| 56,976,657	| 1.023	| 58,287,120 |
|2005	| 56,786,410	| 1.051	| 59,682,517 |
|2006	| 54,641,339	| 1.110	| 60,651,886 |
|2007	| 48,853,563	| 1.292	| 63,118,803 |
|Total	| 543,481,587	| 		| 569,172,456 |

### Estimation of ultimate paid claims
(in progress...)

In [28]:
# Without tail factor
xyz_dev_paid = cl.Pipeline(
    [("dev", cl.Development(average="volume", n_periods=3))]
).fit_transform(xyz_tri['Paid'])
xyz_model_paid = cl.Chainladder().fit(xyz_dev_paid)
xyz_model_paid.ldf_.to_frame().style.format(precision=3)

In [40]:
# With constant tail factor
xyz_dev_paid = cl.Pipeline([
    ("dev", cl.Development(average="volume", n_periods=3)),
    ("tail", cl.TailConstant(1.002))
]
).fit_transform(xyz_tri['Paid'])
xyz_model_paid = cl.Chainladder().fit(xyz_dev_paid)
xyz_model_paid.ldf_.to_frame().style.format(precision=3)

In [42]:
xyz_model_paid.cdf_.to_frame().style.format(precision=3)

In [77]:
xyz_paid = xyz_model_paid.ultimate_.to_frame()
xyz_paid.index = [x.year for x in xyz_paid.index]
xyz_paid.style.format('{:,.0f}')

In [78]:
wiki_paid = pd.DataFrame([47739475, 51204536, 54860424, 56493084, 58963359, 58167880, 58345519, 59963673, 61223522, 65079626], index=range(1998,2008))
wiki_paid.style.format('{:,.0f}')

In [75]:
diff = (xyz_paid['2261']-wiki_paid[0])/wiki_paid[0]
diff.reset_index().style.format({0:'{:.3%}'})